source/sigv4.c

```diff
@@ -793,6 +869,330 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendCanonicalAndSignedHeadersToCanonicalRequest( const char * pHeaders,
+                                                                            size_t headersLen,
+                                                                            uint32_t flags,
+                                                                            CanonicalContext_t * canonicalRequest )
+    {
+        const char * start;
+        const char * end;
+        size_t fieldFlag = 1, noOfHeaders = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        start = end = pHeaders;
+
+        for( i = 0; i < headersLen; i++ )
+        {
+            if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
+            {
+                sigV4Status = SigV4MaxHeaderPairCountExceeded;
+            }
+
+            /* Extracting each header key and value from the headers string. */
+            if( fieldFlag == 1 )
+            {
+                if( pHeaders[ i ] == ':' )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( end - start );
+                    start = end + 1U;
+                    fieldFlag = 0;
+                }
+            }
+            else
+            {
+                if( ( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\r' ) && ( ( i + 1 ) < headersLen ) && ( pHeaders[ i + 1 ] == '\n' ) ) )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                    start = end + 2U;
+                    fieldFlag = 1;
+                    noOfHeaders++;
+                }
+                else if( ( ( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\n' ) ) )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                    start = end + 1U;
+                    fieldFlag = 1;
+                    noOfHeaders++;
+                }
+            }
+
+            end++;
+        }
+
+        /* Sorting headers based on keys. */
+        if( ( sigV4Status == SigV4Success ) && !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+        {
+            qsort( canonicalRequest->pHeadersLoc, noOfHeaders, sizeof( SigV4KeyValuePair_t ), cmpField );
+        }
+
+        if( ( sigV4Status == SigV4Success ) && !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+        {
+            sigV4Status = appendCanonicalizedHeaders( noOfHeaders, canonicalRequest );
+        }
```

Comment:
It doesn't look like `sigV4Status` is changed in the first block so they can be combined? Also, if the `SIGV4_HTTP_PATH_IS_CANONICAL_FLAG` is set then will the canonical headers never be appended?
```suggestion
        if( ( sigV4Status == SigV4Success ) && !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
        {
            qsort( canonicalRequest->pHeadersLoc, noOfHeaders, sizeof( SigV4KeyValuePair_t ), cmpField );
            sigV4Status = appendCanonicalizedHeaders( noOfHeaders, canonicalRequest );
        }
```